src/test/java/edu/hm/hafner/util/SetTest.java

```diff
@@ -0,0 +1,162 @@
+
+        // When
+        assertThat(sut.addAll(Arrays.asList(elements))).isTrue();
+        assertThat(sut.addAll(Arrays.asList(elements))).isFalse();
+        assertThat(sut.add(5)).isTrue();
+        assertThat(sut.add(5)).isFalse();
+
+        // Then
+        assertThat(sut.isEmpty()).isFalse();
+        assertThat(sut.size()).isEqualTo(6);
+        assertThat(sut).containsExactlyInAnyOrder(0, 1, 2, 3, 4, 5);
+        assertThat(sut.contains(0)).isTrue();
+        assertThat(sut.contains(1)).isTrue();
+        assertThat(sut.contains(2)).isTrue();
+        assertThat(sut.contains(3)).isTrue();
+        assertThat(sut.contains(4)).isTrue();
+        assertThat(sut.contains(5)).isTrue();
+    }
+
+    @Test
+    void shouldNotContainRemovedElements() {
+        // Given
+        Integer[] elements = new Integer[] {0, 1, 2, 3, 4};
+        Set<Integer> sut = new HashSet<>(Arrays.asList(elements));
+
+        // When
+        assertThat(sut.remove(0)).isTrue();
+        assertThat(sut.remove(0)).isFalse();
+        assertThat(sut.removeAll(Arrays.asList(1, 2))).isTrue();
+        assertThat(sut.removeAll(Arrays.asList(1, 2))).isFalse();
+
+        // Then
+        assertThat(sut.size()).isEqualTo(2);
+        assertThat(sut).size().isEqualTo(2);
+        assertThat(sut.contains(0)).isFalse();
+        assertThat(sut.contains(1)).isFalse();
+        assertThat(sut.contains(2)).isFalse();
+        assertThat(sut).doesNotContain(0, 1, 2);
+    }
+
+    @Test
+    void shouldRetainElements() {
+        // Given
+        Integer[] elements = new Integer[] {0, 1, 2, 3, 4};
+        Set<Integer> sut = new HashSet<>(Arrays.asList(elements));
+
+        // When
+        assertThat(sut.retainAll(Arrays.asList(3, 4))).isTrue();
+        assertThat(sut.retainAll(Arrays.asList(3, 4))).isFalse();
+
+        // Then
+        assertThat(sut.size()).isEqualTo(2);
+        assertThat(sut).size().isEqualTo(2);
+        assertThat(sut.contains(0)).isFalse();
+        assertThat(sut.contains(1)).isFalse();
+        assertThat(sut.contains(2)).isFalse();
+        assertThat(sut).containsExactlyInAnyOrder(3, 4);
+    }
+
+    @Test
+    void shouldBeEmptyAfterClear() {
+        // Given
+        Integer[] elements = new Integer[] {0, 1, 2, 3, 4};
+        Set<Integer> sut = new HashSet<>(Arrays.asList(elements));
+
+        // When
+        sut.add(5);
+        sut.clear();
+
+        // Then
+        assertThat(sut.isEmpty()).isTrue();
+        assertThat(sut).isEmpty();
+        assertThat(sut.size()).isZero();
+        assertThat(sut).size().isZero();
+    }
+
+    @Test
+    void shouldConvertToArray() {
+        // Given
+        Integer[] elements = new Integer[] {0, 1, 2, 3, 4};
+        Set<Integer> sut = new HashSet<>(Arrays.asList(elements));
+
+        // When
+        Integer[] arr = sut.toArray(new Integer[0]);
+
+        // Then
+        assertThat(arr).isNotNull();
+        assertThat(sut).containsExactlyInAnyOrder(arr);
+    }
+}
```

Comment:
Keine Tests der `clone` und `iterator` Methoden. Auch für die Methode `size` wäre ein eigener Test angebracht.

Reply:
`clone` verstehe ich, aber werden die Iteratoren nicht auch durch JUnit überprüft z.B. bei `containsExactlyInAnyOrder`, `doesNotContain` oder `assertThat(...).size`? Diese Funktionen beziehen sich alle auf Iterables und ich gehe davon aus, dass die Iteratorfuntion damit hinreichend geprüft wird.

Reply:
Ja die entsprechenden Methoden werden dort Verwendet. Ob das aber ausreicht ist nicht direkt ersichtlich und sollte hier auch nicht die Thematik sein. Es war in dieser Aufgabe so angedacht, dass man für jede der HashSet-Methoden mindestens einen Test schreibt, um einfach mal Tests zu schreiben. Wie man herausfindet wie gut ein bestimmter Code getestet wird, wird noch Thema der Vorlesung.